frame/evm/src/lib.rs

```diff
@@ -743,6 +744,7 @@ impl<T: Config> Pallet<T> {
 		nonce: Option<U256>,
 		access_list: Vec<(H160, Vec<H256>)>,
 		is_transactional: bool,
+		omit_fee: bool,
```

Comment:
what is the meaning of this arg? Omit fee on evm layer and pay on substrate layer?

Reply:
Exactly. So as we discussed I first implemented ability to have a different payer for the gas fees (e.g. source=trader, payer=user), but found that 1) it makes testing more difficult because you have to account for gas fees when testing before/after balances and 2) it complicates the final weight calculation.

The gas fee is withdrawn during the execution ([here](https://github.com/moonsama/frontier/blob/master/frame/evm/src/runner/stack.rs#L264)) and then [`PostDispatchInfo`](https://crates.parity.io/frame_support/weights/struct.PostDispatchInfo.html) is used to charge the user only the gas fee, for EVM compatibility (on successful execution `pays_fee=No` so no additional fees are taken). As you can see `PostDispatchInfo` can only reduce fees, not increase them. 

My approach now is to have the user only pay on the substrate layer, which should also reflect weight of evm computation, though I need to test with benchmarks. Alternatively we could also use [this approach](https://docs.substrate.io/build/tx-weights-fees/#post-dispatch-weight-correction).